ledger/src/shred/merkle.rs

```diff
@@ -434,9 +589,133 @@ fn make_merkle_branch(
     Some(MerkleBranch { root, proof })
 }
 
+pub(super) fn recover(mut shreds: Vec<Shred>) -> Result<Vec<Shred>, Error> {
+    // Grab {common, coding} headers from first coding shred.
+    // TODO: Verify that other shreds are consistent!
+    let (mut common_header, mut coding_header) = {
+        let shred = match shreds.iter().find_map(|shred| match shred {
+            Shred::ShredCode(shred) => Some(shred),
+            Shred::ShredData(_) => None,
+        }) {
+            None => return Ok(Vec::default()),
+            Some(shred) => shred,
+        };
+        (*shred.common_header(), *shred.coding_header())
+    };
+    common_header.index -= u32::from(coding_header.position);
+    coding_header.position = 0u16;
+    let proof_size = match common_header.shred_variant {
+        ShredVariant::MerkleCode(proof_size) => proof_size,
+        ShredVariant::MerkleData(_) | ShredVariant::LegacyCode | ShredVariant::LegacyData => {
+            return Err(Error::InvalidShredVariant)
+        }
+    };
+    let num_data_shreds = usize::from(coding_header.num_data_shreds);
+    let num_coding_shreds = usize::from(coding_header.num_coding_shreds);
+    let num_shards = num_data_shreds + num_coding_shreds;
+    // Obtain erasure encoded shards from shreds.
+    let shreds = {
+        let mut batch = vec![None; num_shards];
+        while let Some(shred) = shreds.pop() {
+            let index = match shred.erasure_shard_index() {
+                Ok(index) if index < batch.len() => index,
+                _ => continue,
```

Comment:
should we track out of bounds shred count? or consider this an error?

Reply:
changed to error